textx/scoping/rrel.py

```diff
@@ -368,59 +393,67 @@ def get_from_zero_or_more(obj, lookup_list, matched_path,
+        if self.path_elements[0] == "^":
+            self.path_elements[0] = RRELZeroOrMore(
+                RRELBrackets(RRELSequence([RRELPath([RRELDots(2)])]))
+            )
 
     def __repr__(self):
         if isinstance(self.path_elements[0], RRELDots):
-            return str(self.path_elements[0]) + '.'.join(
-                map(lambda x: str(x), self.path_elements[1:]))
+            return str(self.path_elements[0]) + ".".join(
+                map(lambda x: str(x), self.path_elements[1:])
+            )
         else:
-            return '.'.join(map(lambda x: str(x), self.path_elements))
+            return ".".join(map(lambda x: str(x), self.path_elements))
 
     def start_locally(self):
         return self.path_elements[0].start_locally()
 
     def start_at_root(self):
         return self.path_elements[0].start_at_root()
 
-    def get_next_matches(self, obj, lookup_list, allowed, matched_path,
-                         first_element=False):
+    def get_next_matches(
+        self, obj, lookup_list, allowed, matched_path, first_element=False
+    ):
         from textx.scoping import Postponed
 
-        def intern_get_next_matches(obj, lookup_list, allowed, matched_path,
-                                    first_element=False, idx=0):
+        def intern_get_next_matches(
+            obj, lookup_list, allowed, matched_path, first_element=False, idx=0
+        ):
             assert len(self.path_elements) > idx
             e = self.path_elements[idx]
             for iobj, ilookup_list, imatched_path in e.get_next_matches(
-                    obj, lookup_list, allowed, matched_path, first_element):
+                obj, lookup_list, allowed, matched_path, first_element
+            ):
                 if isinstance(iobj, Postponed):
                     yield iobj, ilookup_list, imatched_path
                     return
                 if len(self.path_elements) - 1 == idx:
                     yield iobj, ilookup_list, imatched_path
                 else:
-                    for iiobj, iilookup_list, \
-                        iimatched_path in intern_get_next_matches(
-                            iobj, ilookup_list, allowed, imatched_path,
-                            first_element=False, idx=idx + 1):
-                        yield iiobj, iilookup_list, iimatched_path
+                    yield from intern_get_next_matches(
+                        iobj,
+                        ilookup_list,
+                        allowed,
+                        imatched_path,
+                        first_element=False,
+                        idx=idx + 1,
+                    )
 
-        for iobj, ilookup_list, imatched_path in intern_get_next_matches(
-                obj, lookup_list, allowed, matched_path, first_element, 0):
-            yield iobj, ilookup_list, imatched_path
+        yield from intern_get_next_matches(
+            obj, lookup_list, allowed, matched_path, first_element, 0
+        )
```

Comment:
If this was done by an automated refactoring; wow!

Reply:
I believe in unittests ;-)

Reply:
> If this was done by an automated refactoring; wow!

I have to disappoint you but no, refactor of this piece is manual but ruff nicely reported the issue. :smile:
ruff is mainly linter/formater but it can do some light refactorings like sorting imports, refactor `str.format` to f-strings etc. which is nice especially when dealing with larger code with a lot of instances of these simple refactorings.